tools/ralph_marketplace_budget.py

```
@@ -0,0 +1,119 @@
+
+Handles iteration budget allocation across smart bots based on weights and efficiency.
+"""
+
+from tools.ralph_marketplace_state import BotState, MarketplaceState
+
+
+def allocate_budgets(state: MarketplaceState) -> dict[str, int]:
+    """Allocate iteration budgets to bots using fixed pool with efficiency rewards.
+
+    The total budget pool is fixed at (num_bots * base_budget_per_bot).
+    Allocation is based on combined score: weight × efficiency.
```

Ruff (0.14.14)

[warning] Docstring contains ambiguous `×` (MULTIPLICATION SIGN). Did you mean `x` (LATIN SMALL LETTER X)?

(RUF002)